code/modules/mining/machine_silo.dm

```diff
@@ -133,31 +163,53 @@ GLOBAL_LIST_EMPTY(silo_access_logs)
 
 	switch(action)
 		if("remove")
-			var/datum/component/remote_materials/remote = locate(params["ref"]) in ore_connected_machines
-			remote?.disconnect_from(src)
+			var/index = params["id"]
+			if(isnull(index))
+				return
+
+			index = text2num(index)
+			if(isnull(index))
+				return
+
+			var/datum/component/remote_materials/remote = ore_connected_machines[index]
+			if(isnull(remote))
+				return
+
+			remote.disconnect_from(src)
 			return TRUE
 
 		if("hold")
-			var/datum/component/remote_materials/remote = locate(params["ref"]) in ore_connected_machines
-			remote?.toggle_holding()
+			var/index = params["id"]
+			if(isnull(index))
+				return
+
+			index = text2num(index)
+			if(isnull(index))
+				return
+
+			var/datum/component/remote_materials/remote = ore_connected_machines[index]
+			if(isnull(remote))
+				return
+
+			remote.toggle_holding()
 			return TRUE
 
 		if("eject")
 			var/datum/material/ejecting = locate(params["ref"])
-			var/amount = text2num(params["amount"])
-			if(!isnum(amount) || !istype(ejecting))
-				return TRUE
+			if(!istype(ejecting))
+				return
+
+			var/amount = params["amount"]
+			if(isnull(amount))
+				return
+
+			amount = text2num(amount)
+			if(isnull(amount))
```

Comment:
Why do you use `isnull` instead of `isnum`?

Reply:
because test2num will already do that check for us internally & if it fails it returns null